# Exploring Hacker News Posts


## Introduction 

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

## Data Columns 

Below are descriptions of the columns:

* id: The unique identifier from Hacker News for the post
* title: The title of the post
* url: The URL that the posts links to, if it the post has a URL
* num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* num_comments: The number of comments that were made on the post
* author: The username of the person who submitted the post
* created_at: The date and time at which the post was submitted

## Objective 

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

- Ask HN: How to improve my personal website?
- Ask HN: Am I the only one outraged by Twitter shutting down share counts?
- Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

- Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
- Show HN: Something pointless I made
- Show HN: Shanhu.io, a programming playground powered by e8vm

We'll compare these two types of posts to determine the following:

1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on average?

## Removing Headers from a List of Lists

First, we should read in the hacker_news.csv file as a list of lists and seperate the header from the rest of the data.

In [4]:
# read in the csv file
from csv import reader
open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)
open_file.close()

# seperate the header and the data
header = hn[0]
data = hn[1:]

# display the first 3 rows and the header
print(header, "\n")
print(data[1], "\n")
print(data[2], "\n")
print(data[3], "\n")

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'] 

['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'] 

['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'] 

['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'] 



## Extracting Ask HN and Show HN Posts

Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

In [10]:
# seperate the posts into 3 categories ask posts, show posts and other posts
ask_posts, show_posts, other_posts = [], [], []

for row in hn:
    title = row[1]
    
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
# check the length of each list to see the quantity of each post type
print("Number of Ask Posts:", len(ask_posts))
print("Number of Show Posts:", len(show_posts))
print("Number of Other Posts:", len(other_posts))

Number of Ask Posts: 1744
Number of Show Posts: 1162
Number of Other Posts: 17195


## Calculating the Average Number of Comments for Ask HN and Show HN Posts 

Here, let us determine if either the ask posts or show posts receive more comments on average.

In [11]:
# find total number of comments in ask posts
total_ask_comments = 0

for post in ask_posts:
    num_comments = int(post[4])
    total_ask_comments += num_comments

# compute average number of comments in ask posts
avg_ask_comments = total_ask_comments / len(ask_posts)

# find total number of comments in show posts
total_show_comments = 0

for post in show_posts:
    num_comments = int(post[4])
    total_show_comments += num_comments
    
# compute average number of comments in show posts
avg_show_comments = total_show_comments / len(show_posts)

# display results
print("Average number of comments in ask posts:", avg_ask_comments)
print("Average number of comments in show posts:", avg_show_comments)

Average number of comments in ask posts: 14.038417431192661
Average number of comments in show posts: 10.31669535283993


<b>Comment:</b> On average, the ask posts receive more comments than the show posts. Thus, we will focus our analysis on the ask posts as they are likely to receive more comments.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. 

## Finding the Amount of Ask Posts and Comments by Hour Created 

We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
2. Calculate the average number of comments ask posts receive by hour created.

In [26]:
import datetime as dt

result_list = []
for post in ask_posts:
    created_at = post[6]
    num_comments = int(post[4])
    result_list.append([created_at, num_comments])

# counts_by_hour: contains the number of ask posts created during each hour of the day.
# comments_by_hour: contains the corresponding number of comments ask posts created at each hour received.
counts_by_hour, comments_by_hour = {}, {}

for item in result_list:
    time = item[0]
    time = dt.datetime.strptime(time, "%m/%d/%Y %H:%M")
    hour = time.strftime("%H")
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = item[1]
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += item[1]

print("Post Count by Hour")
print(counts_by_hour)
print()
print("Comment Frequency by Hour")
print(comments_by_hour)

Post Count by Hour
{'01': 60, '05': 46, '11': 58, '02': 58, '18': 109, '22': 71, '00': 55, '23': 68, '09': 45, '16': 108, '10': 59, '15': 116, '14': 107, '17': 100, '13': 85, '21': 109, '04': 47, '06': 44, '08': 48, '07': 34, '20': 80, '19': 110, '03': 54, '12': 73}

Comment Frequency by Hour
{'01': 683, '05': 464, '11': 641, '02': 1381, '18': 1439, '22': 479, '00': 447, '23': 543, '09': 251, '16': 1814, '10': 793, '15': 4477, '14': 1416, '17': 1146, '13': 1253, '21': 1745, '04': 337, '06': 397, '08': 492, '07': 267, '20': 1722, '19': 1188, '03': 421, '12': 687}


## Calculating the Average Number of Comments for Ask HN Posts by Hour

Now, we will use the two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [29]:
# calculate the average number of comments per post for posts created during each hour
avg_by_hour = []

for hour in counts_by_hour.keys():
    hour_avg = comments_by_hour[hour] / counts_by_hour[hour]
    avg_comments = [hour, hour_avg]
    avg_by_hour.append(avg_comments)
    
# display the list of lists
for item in avg_by_hour:
    print(item)

['01', 11.383333333333333]
['05', 10.08695652173913]
['11', 11.051724137931034]
['02', 23.810344827586206]
['18', 13.20183486238532]
['22', 6.746478873239437]
['00', 8.127272727272727]
['23', 7.985294117647059]
['09', 5.5777777777777775]
['16', 16.796296296296298]
['10', 13.440677966101696]
['15', 38.5948275862069]
['14', 13.233644859813085]
['17', 11.46]
['13', 14.741176470588234]
['21', 16.009174311926607]
['04', 7.170212765957447]
['06', 9.022727272727273]
['08', 10.25]
['07', 7.852941176470588]
['20', 21.525]
['19', 10.8]
['03', 7.796296296296297]
['12', 9.41095890410959]


Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [40]:
# swap the hour with their respective average comments total
swap_avg_by_hour = []
for item in avg_by_hour:
    hour = item[0]
    avg = item[1]
    swap_avg_by_hour.append([avg, hour])
    
# sort the avg comments in descending order
sorted_swap = sorted(swap_avg_by_hour, reverse = True)

print("Top 5 Hours for Ask Posts Comments")
for item in sorted_swap[:5]:
    hour = dt.datetime.strptime(item[1], "%H")
    hour = hour.strftime("%H:%M")
    display_string = "{hour}: {avg:.2f} average comments per post".format(hour=hour, avg=item[0])
    print(display_string)

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


<b>Conclusion:<\b> Thus, the best time to create a ask post that results in the most interaction by other users (i.e. commenting on your post) is during these times. It is clear that the best time to as is at around 3pm as each post gets around 38.59 comments followed by 2 am with 23.81 comments.